File: backend/src/main/scala/slaydemo/backend/battle/objects/BattleWeaponState.scala
package slaydemo.backend.battle.objects

final case class BattleWeaponState(
  weaponKind: WeaponKind,
  ammoInMagazine: AmmoCount,
  magazineSize: AmmoCount,
  reserveAmmo: Option[AmmoCount],
  fireCooldownMs: CooldownMillis,
  reloadRemainingMs: CooldownMillis,
  heat: Int,
  thermalState: BattleWeaponThermalState
) {
  def overheated: Boolean =
    BattleWeaponThermalState.overheatedFlag(thermalState)

  def overheatRemainingMs: CooldownMillis =
    BattleWeaponThermalState.overheatRemainingMs(thermalState)
}

File: frontend/src/objects/battle/microservices/combat/objects/weapon/BattleWeaponState.ts
import type { AmmoCount, BattleWeaponHeat } from "./BattleWeaponScalars";
import type { WeaponKind } from "./WeaponKind";

export interface BattleWeaponState {
  weaponKind: WeaponKind;
  ammoInMagazine: AmmoCount;
  magazineSize: AmmoCount;
  reserveAmmo: AmmoCount | null;
  fireCooldownMs: number;
  reloadRemainingMs: number;
  heat: BattleWeaponHeat;
  overheated: boolean;
  overheatRemainingMs: number;
}

export interface BattleWeaponInventoryState {
  currentWeaponIndex: number;
  weapons: BattleWeaponState[];
}

